tests/store/test_sqlalchemy_store.py

```diff
@@ -498,3 +507,173 @@ def test_restore_run(self):
+
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ALL))
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ACTIVE_ONLY))
+        self.assertSequenceEqual([], self._search(exp, run_view_type=ViewType.DELETED_ONLY))
+
+        first = runs[0]
+
+        self.store.delete_run(first)
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ALL))
+        self.assertSequenceEqual(runs[1:], self._search(exp, run_view_type=ViewType.ACTIVE_ONLY))
+        self.assertSequenceEqual([first], self._search(exp, run_view_type=ViewType.DELETED_ONLY))
+
+        self.store.restore_run(first)
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ALL))
+        self.assertSequenceEqual(runs, self._search(exp, run_view_type=ViewType.ACTIVE_ONLY))
+        self.assertSequenceEqual([], self._search(exp, run_view_type=ViewType.DELETED_ONLY))
+
+    def test_search_params(self):
+        experiment_id = self._experiment_factory('search_params').experiment_id
+        r1 = self._run_factory('r1', experiment_id).to_mlflow_entity().info.run_uuid
+        r2 = self._run_factory('r2', experiment_id).to_mlflow_entity().info.run_uuid
+
+        self.store.log_param(r1, entities.Param('generic_param', 'p_val'))
+        self.store.log_param(r2, entities.Param('generic_param', 'p_val'))
+
+        self.store.log_param(r1, entities.Param('p_a', 'abc'))
+        self.store.log_param(r2, entities.Param('p_b', 'ABC'))
+
+        expr = self._param_expression("generic_param", "=", "p_val")
+        self.assertSequenceEqual([r1, r2], self._search(experiment_id, param_expressions=[expr]))
+
+        expr = self._param_expression("generic_param", "=", "wrong_val")
+        self.assertSequenceEqual([], self._search(experiment_id, param_expressions=[expr]))
+
+        expr = self._param_expression("generic_param", "!=", "p_val")
+        self.assertSequenceEqual([], self._search(experiment_id, param_expressions=[expr]))
+
+        expr = self._param_expression("generic_param", "!=", "wrong_val")
+        self.assertSequenceEqual([r1, r2], self._search(experiment_id, param_expressions=[expr]))
+
+        expr = self._param_expression("p_a", "=", "abc")
+        self.assertSequenceEqual([r1], self._search(experiment_id, param_expressions=[expr]))
+
+        expr = self._param_expression("p_b", "=", "ABC")
+        self.assertSequenceEqual([r2], self._search(experiment_id, param_expressions=[expr]))
+
+    def test_search_metrics(self):
+        experiment_id = self._experiment_factory('search_params').experiment_id
+        r1 = self._run_factory('r1', experiment_id).to_mlflow_entity().info.run_uuid
+        r2 = self._run_factory('r2', experiment_id).to_mlflow_entity().info.run_uuid
+
+        self.store.log_metric(r1, entities.Metric("common", 1.0, 1))
+        self.store.log_metric(r2, entities.Metric("common", 1.0, 1))
```

Comment:
Let's make sure to include a test with the same metric name across multiple runs with different timestamps. We've had a bug in the past with that particular setup.

Reply:
That's a good call out.